### PR TITLE
Add support for Semantic Version label postfixes.

### PR DESCRIPTION
**Pull request by mossblaser**
Adds a new field to the CoreInfo and BMPInfo namedtuples which contains any
version postfixes (e.g. -dev or build numbers).

Also refactors version-number parsing so that the implementation is shared
between BMPController and MachineController.

### CODE REVIEW

**Comment by mossblaser** [comment]
What on earth is going on with the tests :s can't reproduce locally atm... :s

**Comment by mossblaser** [comment]
Ah... helps if you add new files... /me is an idiot as usual...

**Comment by mossblaser** [comment]
(And this is why continuous integration is good :P)

File: rig/machine_control/bmp_controller.py [hunk 1]
```diff
@@ -416,7 +408,7 @@ class BMPInfo(collections.namedtuple(
         The position of the board in a frame. (This should correspond exactly
         with a board's board-coordinate.
     version : (major, minor, patch)
-        Software version number.
+        Software version number. See also: ``software_version_labels``.
```
Comment (mundya) [inline rig/machine_control/bmp_controller.py]: Where is `software_version_labels`? Don't need the `software`?

File: rig/machine_control/machine_controller.py [hunk 2]
```diff
@@ -474,23 +475,14 @@ def get_software_version(self, x=255, y=255, processor=0):
         pcpu = (sver.arg1 >> 8) & 0xff
         vcpu = sver.arg1 & 0xff
 
-        # arg2 => version number and buffer size
-        version = (sver.arg2 >> 16)
+        # arg2 => version number (parsed seperately) and buffer size
```
Comment (mundya) [inline rig/machine_control/machine_controller.py]: sp. separately

File: rig/machine_control/bmp_controller.py (hunk 1 above)
Comment (mossblaser) [inline rig/machine_control/bmp_controller.py]: Whoops... Spot the copy-paste!

File: rig/machine_control/common.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+"""Internal-use functions shared between the
+:py:class:`~rig.machine_control.MachineController` and
+:py:class:`~rig.machine_control.BMPController` objects.
+"""
+
+import re
+
+
+VERSION_NUMBER_REGEX = re.compile(r"^(\d+)[.](\d+)[.](\d+)(\D.*)?$")
```
Comment (mundya) [inline rig/machine_control/common.py]: Is it not fair to assume that the tags will begin with a "-", that is the spec I think?

Comment (mossblaser) [inline rig/machine_control/common.py]: Not quite. For example "1.2.3+9fb2b01" is valid, as is "1.2.3-dev+9fb2b01"

File: rig/machine_control/common.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        minor = legacy_version_field % 100
+        patch = 0
+        labels = ""
+    else:
+        # Semantic Version encoding: packed after the null-terminator of the
+        # software name in the version string.
+        software_name, _, version_number = software_name.partition("\0")
+
+        match = VERSION_NUMBER_REGEX.match(version_number.rstrip("\0"))
+        assert match, "Malformed version number: {}".format(version_number)
```
Comment (mundya) [inline rig/machine_control/common.py]: Is it worth testing that this `assert` is present/replacing it with a more formal error?

File: rig/machine_control/common.py (hunk 3 above)
Comment (mundya) [inline rig/machine_control/common.py]: Cool, thanks

File: rig/machine_control/common.py (hunk 4 above)
Comment (mossblaser) [inline rig/machine_control/common.py]: I'm split on this one. Usually yes but in this case a malformed version probably indicates something really bad is going on so an assert suits fine. The assert is only there to make the explosion more useful than "None has no .group()".

**Comment by mundya** [comment]
Looks good to me apart from the copy-paste and the spelling of "separately".

**Pull request merged**